rs/registry/canister/src/registry.rs

```diff
@@ -284,30 +328,57 @@ impl Registry {
         self.apply_mutations_as_version(mutations, self.version);
     }
 
-    /// Verifies the implicit precondition corresponding to the mutation_type
-    /// field.
+    /// Some mutation_type(s) require that the key/value is currently (i.e.
+    /// right before the mutation is performed) present, while other(s) require
+    /// that the key/value is currently absent. This enforces those requirements.
     fn verify_mutation_type(&self, mutations: &[RegistryMutation]) -> Vec<Error> {
         mutations
             .iter()
-            .map(|m| {
-                let key = &m.key;
-                let latest = self
+            .map(|new_mutation| {
+                let RegistryMutation {
+                    key,
+                    mutation_type,
+                    value: _,
+                } = new_mutation;
+
+                let mutation_type = Type::try_from(*mutation_type).map_err(|err| {
+                    Error::MalformedMessage(format!("Unable to convert mutation_type: {}", err,))
+                })?;
+                let Some(requires_already_present) = mutation_type.requires_already_present()
+                else {
+                    return Ok(());
+                };
+
+                let already_present = self
                     .get_last(key)
-                    .filter(|registry_value| !registry_value.deletion_marker);
-                match (Type::try_from(m.mutation_type).ok(), latest) {
-                    (None, _) => Err(Error::MalformedMessage(format!(
-                        "Unknown mutation type {} for key {:?}.",
-                        m.mutation_type, m.key
-                    ))),
-                    (Some(Type::Insert), None) => Ok(()),
-                    (Some(Type::Insert), Some(_)) => Err(Error::KeyAlreadyPresent(key.to_vec())),
-                    (Some(Type::Update), None) => Err(Error::KeyNotPresent(key.to_vec())),
-                    (Some(Type::Update), Some(_)) => Ok(()),
-                    (Some(Type::Delete), None) => Err(Error::KeyNotPresent(key.to_vec())),
-                    (Some(Type::Delete), Some(_)) => Ok(()),
-                    (Some(Type::Upsert), None) => Ok(()),
-                    (Some(Type::Upsert), Some(_)) => Ok(()),
+                    .map(|registry_value| match registry_value.content {
+                        Some(high_capacity_registry_value::Content::DeletionMarker(
+                            deletion_marker,
+                        )) => !deletion_marker,
+
+                        None
+                        | Some(high_capacity_registry_value::Content::Value(_))
+                        | Some(high_capacity_registry_value::Content::LargeValueChunkKeys(_)) => {
+                            true
+                        }
+                    })
+                    .unwrap_or(false);
```

Comment:
Nit. I suspect clippy will force you to write it like this:

```suggestion
                    .unwrap_or_default();
```

Reply:
I see no such complaints.